host_core/lib/host_core/providers/provider_supervisor.ex

```diff
@@ -193,113 +162,164 @@ defmodule HostCore.Providers.ProviderSupervisor do
-
-          prefix = HostCore.Host.lattice_prefix()
-
-          # Allow provider 2 seconds to respond/acknowledge termination request (give time to clean up resources)
-          case HostCore.Nats.safe_req(
-                 :lattice_nats,
-                 "wasmbus.rpc.#{prefix}.#{public_key}.#{link_name}.shutdown",
-                 Jason.encode!(%{host_id: HostCore.Host.host_key()}),
-                 receive_timeout: 2000
-               ) do
-            {:ok, _msg} ->
-              :ok
-
-            {:error, :no_responders} ->
-              Logger.error("No responders to request to terminate provider")
-              :error
-
-            {:error, :timeout} ->
-              Logger.error("No capability providers responded to shutdown request")
-              :error
-          end
-
-          # Pause for n milliseconds between shutdown request and forceful termination
-          Process.sleep(HostCore.Host.provider_shutdown_delay())
-          ProviderModule.halt(pid)
-
-        [] ->
-          Logger.warn(
-            "No provider is running with public key #{public_key} and link name \"#{link_name}\"",
-            provider_id: public_key,
-            link_name: link_name
-          )
+  defp get_reference_key(reference, lattice_prefix) do
+    lookup_reference_key(reference, lattice_prefix)
+  end
+
+  defp lookup_reference_key(reference, lattice_prefix) do
+    case HostCore.Refmaps.Manager.lookup_refmap(lattice_prefix, reference) do
+      {:ok, {_oci, pk}} -> pk
+      _ -> ""
+    end
+  end
+
+  defp is_running?(key, _, _) when byte_size(key) == 0, do: false
+
+  defp is_running?(key, link_name, host_id) do
+    Enum.any?(
+      all_providers(host_id),
+      fn {_pid, public_key, ln, _contract_id, _instance_id} ->
+        public_key == key && link_name == ln
       end
+    )
+  end
+
+  def terminate_provider(host_id, public_key, link_name) do
+    for {pid, pk, link, _contract_id, _instance_id} <- all_providers(host_id),
+        pk == public_key,
+        link == link_name do
+      Logger.info("About to terminate provider process",
+        provider_id: public_key,
+        link_name: link_name
+      )
+
+      ProviderModule.halt(pid)
     end
   end
 
-  def terminate_all() do
-    all_providers()
-    |> Enum.each(fn {_pid, pk, link, _contract, _instance_id} -> terminate_provider(pk, link) end)
+  def terminate_all(host_id) when is_binary(host_id) do
+    for {pid, _pk, _link, _contract, _instance_id} <- all_providers(host_id) do
+      ProviderModule.halt(pid)
+    end
   end
 
   @doc """
   Produces a list of tuples in the form of {pid, public_key, link_name, contract_id, instance_id}
   of all of the current providers running
   """
-  def all_providers() do
-    Supervisor.which_children(HostCore.Providers.ProviderSupervisor)
-    |> Enum.map(fn {_d, pid, _type, _modules} ->
-      provider_for_pid(pid)
+  def all_providers(host_id) do
+    # $1 - {pk, link_name}
+    # $2 - pid
+    # $3 - host_id
```

Comment:
Is this the right shape? It looks like it's a list of `{{pk, link_name}, pid}` but I might be wrong